moped-editor/src/views/projects/projectView/ProjectPhase/helpers.js

```diff
@@ -0,0 +1,183 @@
+  phase_start: yup
+    .date()
+    .nullable()
+    .optional()
+    .when("is_current_phase", {
+      is: true,
+      then: (schema) =>
+        schema.required("Start date is required when phase is current"),
+    })
+    .typeError("Invalid Date"),
+  phase_end: yup.date().nullable().optional().typeError("Invalid Date"),
+  is_current_phase: yup.boolean(),
+  is_phase_start_confirmed: yup.boolean(),
+  is_phase_end_confirmed: yup.boolean(),
+  description: yup
+    .string()
+    .max(500, "Must be less than 500 characters")
+    .nullable(),
+  project_phase_id: yup.number().nullable().optional(),
+  project_id: yup.number().required(),
+});
+
+const DEFAULT_FORM_VALUES = {
+  project_phase_id: null,
+  phase_id: null,
+  subphase_id: null,
+  phase_start: null,
+  is_phase_start_confirmed: false,
+  phase_end: null,
+  is_phase_end_confirmed: false,
+  phase_description: null,
+  is_current_phase: false,
+  project_id: null,
+};
+
+/**
+ * Hook which provides initial form values
+ * @param {object} phase - an optoinal `moped_proj_phase` object whose values will
+ * override the DEFAULT_FORM_VALUES
+ */
+export const useDefaultValues = (phase) =>
+  useMemo(() => {
+    // initialize form with default values plus the project id
+    let defaultValues = {
+      ...DEFAULT_FORM_VALUES,
+      project_id: phase.project_id,
+    };
+
+    if (phase.project_phase_id) {
+      // we are editing a phase: update all defaults from phase
+      Object.keys(DEFAULT_FORM_VALUES).forEach((key) => {
+        defaultValues[key] = phase[key];
+      });
+    }
+    return defaultValues;
+  }, [phase]);
+
+/**
+ * Hook which returns an array of subphase options given an input `phase_id`
+ * and an array of `moped_phases` objects
+ */
+export const useSubphases = (phase_id, phases) =>
+  useMemo(
+    () =>
+      phase_id
+        ? phases.find((p) => p.phase_id === phase_id)?.moped_subphases || []
+        : [],
+    [phase_id, phases]
+  );
+
+/**
+ * Hook which returns an object of phase IDs with their name. Taking
+ * the shape of  { [phase_id]: phase_name }
+ */
+export const usePhaseNameLookup = (phases) =>
+  useMemo(
+    () =>
+      phases.reduce(
+        (obj, item) =>
+          Object.assign(obj, {
+            [item.phase_id]: item.phase_name,
+          }),
+        {}
+      ),
+    [phases]
+  );
+
+/**
+ * Hook which returns an object of subphase IDs with their name. Taking
+ * the shape of  { [subphase_id]: subphase_name }
+ */
+export const useSubphaseNameLookup = (subphases) =>
+  useMemo(
+    () =>
+      subphases.reduce(
+        (obj, item) =>
+          Object.assign(obj, {
+            [item.subphase_id]: item.subphase_name,
+          }),
+        {}
+      ),
+    [subphases]
+  );
+
+/**
+ * Hoolk which returns an array of project_phase_ids of the project's current phase(s).
+ * Although only one phase should ever be current, we handle the possibilty that there
+ * are multiple
```

Comment:
just curious, why do we handle that possibility? is it because its easier to work with an Array?

Reply:
well, since we don't enforce this constraint at the DB level, my thinking was that it is technically possible for a project to have multiple current phases as result of some busted ETL or data update that happens outside of the Editor.

we could—and maybe should?—create a DB trigger that manages this instead 🤔